### PR TITLE
 Update audioclient & add audiopolicy

### PR DESCRIPTION
**Pull request by trmcnvn**
This PR updates `um::audioclient` to add missing elements and adds a new `um::audiopolicy` feature.

Closes an older PR #645

### CODE REVIEW

**Comment by retep998** [comment]
PR needs to be rebased due to failing CI.

File: src/um/audioclient.rs [hunk 1]
```diff
@@ -51,21 +51,68 @@ pub const AUDCLNT_E_RAW_MODE_UNSUPPORTED: HRESULT = AUDCLNT_ERR!(0x027);
 pub const AUDCLNT_S_BUFFER_EMPTY: SCODE = AUDCLNT_SUCCESS!(0x001);
 pub const AUDCLNT_S_THREAD_ALREADY_REGISTERED: SCODE = AUDCLNT_SUCCESS!(0x002);
 pub const AUDCLNT_S_POSITION_STALLED: SCODE = AUDCLNT_SUCCESS!(0x003);
+pub const AUDCLNT_S_NO_SINGLE_PROCESS: SCODE = AUDCLNT_SUCCESS!(0x00d);
```
Comment (retep998) [inline src/um/audioclient.rs:54]: Where is this constant coming from? I can't find it in version 10.0.18362.0 of the Windows SDK.

File: src/um/audioclient.rs [hunk 2]
```diff
@@ -51,21 +51,68 @@ pub const AUDCLNT_E_RAW_MODE_UNSUPPORTED: HRESULT = AUDCLNT_ERR!(0x027);
...
+}}
+ENUM!{enum AMBISONICS_TYPE {
+    AMBISONICS_TYPE_FULL3D = 0,
+}}
+ENUM!{enum AMBISONICS_CHANNEL_ORDERING {
+    AMBISONICS_CHANNEL_ORDERING_ACN = 0,
+}}
+ENUM!{enum AMBISONICS_NORMALIZATION {
+    AMBISONICS_NORMALIZATION_SN3D = 0,
+    AMBISONICS_NORMALIZATION_N3D = 0x1,
```
Comment (retep998) [inline src/um/audioclient.rs:80]:
```suggestion
    AMBISONICS_NORMALIZATION_N3D,
```

File: src/um/audiopolicy.rs [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+use um::audiosessiontypes::AudioSessionState;
+use um::unknwnbase::{IUnknown, IUnknownVtbl};
+use um::winnt::{HRESULT, LPCWSTR, LPWSTR};
+ENUM!{enum AudioSessionDisconnectReason {
+  DisconnectReasonDeviceRemoval = 0,
+  DisconnectReasonServerShutdown = 0x1,
+  DisconnectReasonFormatChanged = 0x2,
+  DisconnectReasonSessionLogoff = 0x3,
+  DisconnectReasonSessionDisconnected = 0x4,
+  DisconnectReasonExclusiveModeOverride = 0x5,
```
Comment (retep998) [inline src/um/audiopolicy.rs:15]:
```suggestion
  DisconnectReasonServerShutdown,
  DisconnectReasonFormatChanged,
  DisconnectReasonSessionLogoff,
  DisconnectReasonSessionDisconnected,
  DisconnectReasonExclusiveModeOverride,
```

File: src/um/audiopolicy.rs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    NewDisplayName: LPCWSTR,
+    EventContext: LPCGUID,
+  ) -> HRESULT,
+  fn OnIconPathChanged(
+    NewIconPath: LPCWSTR,
+    EventContext: LPCGUID,
+  ) -> HRESULT,
+  fn OnSimpleVolumeChanged(
+    NewVolume: c_float,
+    NewMute: BOOL,
```
Comment (retep998) [inline src/um/audiopolicy.rs:45]:
```suggestion
    NewMute: BOOL,
    EventContext: LPCGUID,
```

**Review by retep998 — CHANGES_REQUESTED** [review]

File: src/um/audioclient.rs (hunk 1 above)
Comment (trmcnvn) [reply src/um/audioclient.rs:54]: https://docs.microsoft.com/en-us/windows/win32/api/audiopolicy/nf-audiopolicy-iaudiosessioncontrol2-getprocessid

```
The session spans more than one process. In this case, pRetVal receives the initial identifier of the process that created the session. To use this value , include the following definition:

#define AUDCLNT_S_NO_SINGLE_PROCESS AUDCLNT_SUCCESS (0x00d)
```

**Comment by trmcnvn** [comment]
I'm going to close this PR as #876 supersedes this one. The latest review comments should most likely be addressed there if they haven't already. /cc @JoshuaBatty

**Pull request closed without merging**